src/uthash.h

```diff
@@ -105,6 +102,30 @@ typedef unsigned char uint8_t;
 #define uthash_expand_fyi(tbl)            /* can be defined to log expands   */
 #endif
 
+#ifndef HASH_NOMEM_OK
+#define HASH_NOMEM_OK 0
+#endif
+
+#if HASH_NOMEM_OK
+/* malloc failures can be recovered from */
+
+#ifndef uthash_fatal
+#define uthash_fatal(obj) do{}while(0)
```

Comment:
> The uthash_oom [you mean the thing called uthash_fatal in the current patch, right?] ...

Yes
> Can you walk me through how it would get called, in the HASH_NOMEM_OK=1 case

First, there is my bad; I forgot to rename `hash_mem_failed` back to `hash_fatal`, that's probably what's caused the confusion. And the test didn't catch that because it used the old name too. This PR has been updated with the corresponding changes.

The calls are in `HASH_SELECT` and `HASH_ADD_TO_TABLE`. There are invoked at the exit points of the operations, if it is apparent (`mem_ok` is 0) that adding the element failed.

> if the user called `HASH_ADD_KEYPTR_BYHASHVALUE_INORDER` and the failure happened, let's say, while allocating the initial bloom filter in `HASH_BLOOM_MAKE`?

For bloom allocation to fail, this must be a new table. So, `HASH_MAKE_TABLE` will be called, when bloom is attempted allocation, and failed, _mem_ok will be reset to 0 (provided by `...INORDER`), buckets and table will be freed, and execution will return to `...INORDER`.

The list operations are skipped (because the table is new), and `HASH_ADD_TO_TABLE` is called. The `_mem_ok` is passed there. The `HASH_ADD_TO_TABLE` only executes the actual operations if passed in `mem_ok` is 1, otherwise the element has its table value set to 0, and `uthash_fatal` called on that element. The hash is checked, and `HASH_ADD_TO_TABLE` returns. Note that the same will happen (element's table reset to 0, and `uthash_fatal` called) if the bucket extension fails, this is handled in the other branch of the top-level `if` of `HASH_ADD_TO_TABLE`.